File: backend/src/captionovo/pipeline/orchestrator.py
from captionovo.domain.processing import PipelineJobPayload, build_pipeline_stages
from captionovo.pipeline.context import (
    STAGE_RUNNERS,
    PipelineContext,
    create_pipeline_context,
    dispose_pipeline_context,
    emit_stage_event,
)


class PipelineOrchestrator:
    def __init__(self, *, admin, providers, credits, events) -> None:
        self._admin = admin
        self._providers = providers
        self._credits = credits
        self._events = events

    async def run(self, payload: PipelineJobPayload, job_id: str | None = None) -> None:
        ctx = await create_pipeline_context(
            {
                "admin": self._admin,
                "payload": payload,
                "providers": self._providers,
                "credits": self._credits,
                "events": self._events,
                "job_id": job_id,
            }
        )

        try:
            for step_name in build_pipeline_stages(payload.outputs):
                runner_info = STAGE_RUNNERS.get(step_name)
                if not runner_info:
                    continue
                run_fn, critical, should_run = runner_info
                if not should_run(ctx):
                    continue

                try:
                    await run_fn(ctx)
                except Exception as exc:
                    message = str(exc) if str(exc) else "Stage failed"
                    await emit_stage_event(ctx, step_name, "failed", message)
                    if critical:
                        raise
                    ctx.failed_stages.append(step_name)

            await self._mark_completed(ctx, payload)
        except Exception as exc:
            await self._mark_failed(ctx, payload, exc)
            raise
        finally:
            await dispose_pipeline_context(ctx)

    async def _mark_completed(self, ctx: PipelineContext, payload: PipelineJobPayload) -> None:
        is_partial = len(ctx.failed_stages) > 0

        self._admin.table("projects").update(
            {"processing_state": "completed", "status": "partial" if is_partial else "completed"}
        ).eq("id", payload.project_id).execute()

        if ctx.has_transcript:
            self._credits.commit_from_reservation(payload.project_id, payload.user_id)
        else:
            self._credits.release(payload.project_id, payload.user_id)

        message = (
            f"Partial completion: {', '.join(ctx.failed_stages)}"
            if is_partial
            else "All stages completed"
        )
        ctx.events.emit(
            project_id=payload.project_id,
            job_id=ctx.job_id,
            stage="pipeline",
            status="completed",
            message=message,
        )

    async def _mark_failed(self, ctx: PipelineContext, payload: PipelineJobPayload, error: Exception) -> None:
        message = str(error) if str(error) else "Processing failed"

        self._admin.table("projects").update(
            {"processing_state": "failed", "status": "failed"}
        ).eq("id", payload.project_id).execute()

        if ctx.has_transcript:
            self._admin.table("projects").update(
                {"status": "partial", "processing_state": "transcript_ready"}
            ).eq("id", payload.project_id).execute()
            self._credits.commit_from_reservation(payload.project_id, payload.user_id)
        else:
            self._credits.release(payload.project_id, payload.user_id)

        ctx.events.emit(
            project_id=payload.project_id,
            job_id=ctx.job_id,
            stage="pipeline",
            status="failed",
            message=message,
        )
        print(f"[pipeline] project {payload.project_id} failed: {message}")

File: backend/src/pipeline/orchestrator.ts
import type { PipelineJobPayload } from "../domain/processing.js";
import {
  buildPipelineStages,
  createPipelineContext,
  disposePipelineContext,
  type PipelineContext,
} from "./context.js";
import { pipelineStages } from "./stages/index.js";
import type { ProviderRegistry } from "../providers/types.js";
import type { AppSupabaseClient } from "../lib/supabase.js";
import type { Env } from "../lib/env.js";
import { CreditsService } from "../services/credits.js";
import { ProcessingEventsService } from "../services/processing-events.js";
import { emitStageEvent } from "./context.js";

export interface PipelineOrchestratorDeps {
  env: Env;
  admin: AppSupabaseClient;
  providers: ProviderRegistry;
  credits: CreditsService;
  events: ProcessingEventsService;
}

export class PipelineOrchestrator {
  constructor(private readonly deps: PipelineOrchestratorDeps) {}

  async run(payload: PipelineJobPayload, jobId?: string) {
    const ctx = await createPipelineContext({ ...this.deps, payload, jobId });
    const plannedSteps = buildPipelineStages(payload.outputs);

    try {
      for (const stepName of plannedSteps) {
        const stage = pipelineStages.find((candidate) => candidate.name === stepName);
        if (!stage || !stage.shouldRun(ctx)) continue;

        try {
          await stage.run(ctx);
        } catch (error) {
          const message = error instanceof Error ? error.message : "Stage failed";
          await emitStageEvent(ctx, stage.name, "failed", message);

          if (stage.critical) {
            throw error;
          }

          ctx.failedStages.push(stage.name);
        }
      }

      await this.markCompleted(ctx, payload);
    } catch (error) {
      await this.markFailed(ctx, payload, error);
      throw error;
    } finally {
      await disposePipelineContext(ctx);
    }
  }

  private async markCompleted(ctx: PipelineContext, payload: PipelineJobPayload) {
    const isPartial = ctx.failedStages.length > 0;

    await ctx.admin
      .from("projects")
      .update({
        processing_state: "completed",
        status: isPartial ? "partial" : "completed",
      })
      .eq("id", payload.projectId);

    if (ctx.hasTranscript) {
      await ctx.credits.commitFromReservation(payload.projectId, payload.userId);
    } else {
      await ctx.credits.release(payload.projectId, payload.userId);
    }

    await ctx.events.emit({
      projectId: payload.projectId,
      jobId: ctx.jobId,
      stage: "pipeline",
      status: "completed",
      message: isPartial ? `Partial completion: ${ctx.failedStages.join(", ")}` : "All stages completed",
    });
  }

  private async markFailed(ctx: PipelineContext, payload: PipelineJobPayload, error: unknown) {
    const message = error instanceof Error ? error.message : "Processing failed";

    await ctx.admin
      .from("projects")
      .update({
        processing_state: "failed",
        status: "failed",
      })
      .eq("id", payload.projectId);

    if (ctx.hasTranscript) {
      await ctx.admin
        .from("projects")
        .update({ status: "partial", processing_state: "transcript_ready" })
        .eq("id", payload.projectId);
      await ctx.credits.commitFromReservation(payload.projectId, payload.userId);
    } else {
      await ctx.credits.release(payload.projectId, payload.userId);
    }

    await ctx.events.emit({
      projectId: payload.projectId,
      jobId: ctx.jobId,
      stage: "pipeline",
      status: "failed",
      message,
    });

    console.error(`[pipeline] project ${payload.projectId} failed:`, message);
  }
}
